stan/math/prim/prob/beta_neg_binomial_lpmf.hpp

```diff
@@ -0,0 +1,219 @@
+    }
+  }
+
+  VectorBuilder<!is_constant_all<T_alpha, T_beta>::value, T_partials_return,
+                T_alpha, T_beta>
+      digamma_alpha_beta(size_alpha_beta);
+  if (!is_constant_all<T_alpha, T_beta>::value) {
+    for (size_t i = 0; i < size_alpha_beta; i++) {
+      digamma_alpha_beta[i] = digamma(alpha_vec.val(i) + beta_vec.val(i));
+    }
+  }
+
+  VectorBuilder<!is_constant_all<T_r>::value, T_partials_return, T_n, T_r>
+      digamma_n_r(size_n_r);
+  if (!is_constant_all<T_r>::value) {
+    for (size_t i = 0; i < size_n_r; i++) {
+      digamma_n_r[i] = digamma(n_vec[i] + r_vec.val(i));
+    }
+  }
+
+  VectorBuilder<!is_constant_all<T_r, T_alpha>::value, T_partials_return, T_r,
+                T_alpha>
+      digamma_r_alpha(size_r_alpha);
+  if (!is_constant_all<T_r, T_alpha>::value) {
+    for (size_t i = 0; i < size_r_alpha; i++) {
+      digamma_r_alpha[i] = digamma(r_vec.val(i) + alpha_vec.val(i));
+    }
+  }
+
+  VectorBuilder<!is_constant_all<T_beta>::value, T_partials_return, T_n, T_beta>
+      digamma_n_beta(size_n_beta);
+  if (!is_constant_all<T_n, T_beta>::value) {
+    for (size_t i = 0; i < size_n_beta; i++) {
+      digamma_n_beta[i] = digamma(n_vec[i] + beta_vec.val(i));
+    }
+  }
+
+  VectorBuilder<!is_constant_all<T_r>::value, T_partials_return, T_r> digamma_r(
+      size_r);
+  if (!is_constant_all<T_r>::value) {
+    for (size_t i = 0; i < size_r; i++) {
+      digamma_r[i] = digamma(r_vec.val(i));
+    }
+  }
+
+  VectorBuilder<!is_constant_all<T_alpha>::value, T_partials_return, T_alpha>
+      digamma_alpha(size_alpha);
+  if (!is_constant_all<T_alpha>::value) {
+    for (size_t i = 0; i < size_alpha; i++) {
+      digamma_alpha[i] = digamma(alpha_vec.val(i));
+    }
+  }
+
+  VectorBuilder<!is_constant_all<T_beta>::value, T_partials_return, T_beta>
+      digamma_beta(size_beta);
+  if (!is_constant_all<T_beta>::value) {
+    for (size_t i = 0; i < size_beta; i++) {
+      digamma_beta[i] = digamma(beta_vec.val(i));
+    }
+  }
+
+  for (size_t i = 0; i < max_size_seq_view; i++) {
+    if (include_summand<propto>::value)
+      logp += normalizing_constant[i];
+    logp += lbeta_diff[i];
+
+    if (!is_constant_all<T_r>::value) {
+      partials<0>(ops_partials)[i] += digamma_n_r[i] - digamma_n_r_alpha_beta[i]
+                                      - (digamma_r[i] - digamma_r_alpha[i]);
+    }
+    if (!is_constant_all<T_alpha>::value) {
+      partials<1>(ops_partials)[i] += digamma_alpha_beta[i]
+                                      - digamma_n_r_alpha_beta[i]
+                                      - (digamma_alpha[i] - digamma_r_alpha[i]);
+    }
+    if (!is_constant_all<T_beta>::value) {
+      partials<2>(ops_partials)[i] += digamma_alpha_beta[i]
+                                      - digamma_n_r_alpha_beta[i]
+                                      + digamma_n_beta[i] - digamma_beta[i];
+    }
+  }
+  return ops_partials.build(logp);
```

Comment:
Can you move all of the for loops you have separated in the above into this for loop? This would be nice because it would cut down on a bit of the code and calculations you have to do. For instance `digamma_n_r` and `digamma_r` are only used when `T_r` is an autodiff type. So you can make this like 

```c++
    if (!is_constant_all<T_r>::value) {
      partials<0>(ops_partials)[i] += digamma(n_vec[i] + r_vec.val(i)) - digamma_n_r_alpha_beta[i]
                                      - (digamma(r_vec.val(i)) - digamma_r_alpha[i]);
    }
```

And you only need to calculate `digamma_r_alpha` if `T_r` or `T_alpha` is not a constant

Reply:
Changes look great! Going to rubber duck this on Monday then should be good!